File: archived/MAY15/CHEFCK.cc
#include <bits/stdc++.h>
typedef long long LL;
using namespace std;
const int MAXN = 10000000 + 10, MOD = 1e9 + 7;
LL A[MAXN], ms[MAXN], mp[MAXN];
LL a,b, c, d, e, f, r, s, t, m;
LL L1, La, Lc, Lm, D1, Da, Dc, Dm;
int N, Q, K;

int main() {
  cin >> N >> K >> Q;
  cin >> a >> b >> c >> d >> e >> f >> r >> s >> t >> m;
  cin >> A[0];
  cin >> L1 >> La >> Lc >> Lm >> D1 >> Da >> Dc >> Dm;
  LL pt = t;
  for (int x = 1; x < N; ++ x) {
    pt = pt * t % s;
    if (pt <= r) A[x] = (A[x - 1] * (a * A[x - 1] % m + b) + c) % m;
    else A[x] = (A[x - 1] * (d * A[x - 1] % m + e) + f) % m;
  }
  for (int i = 0; i < N; ++ i) {
    if (i % K == 0) mp[i] = A[i];
    else mp[i] = min(mp[i - 1], A[i]);
  }
  for (int i = N - 1; i >= 0; -- i) {
    if (i % K == K - 1) ms[i] = A[i];
    else ms[i] = min(ms[i + 1], A[i]);
  }
  LL sum = 0, pro = 1;
  for (int i = 1; i <= Q; ++ i) {
    L1 = (La * L1 + Lc) % Lm;
    D1 = (Da * D1 + Dc) % Dm;
    int L = L1 + 1, R = min(L + K - 1 + (int)D1, N);
    -- L; -- R;
    int pl = L / K, pr = R / K, mx;
    if (L % K == 0) mx = min(mp[R], ms[L]);
    else {
      mx = min(ms[L], mp[R]);
      if (pl + 1 != pr) mx = min((LL)mx, mp[pr * K - 1]);
    }
    sum += mx, pro = pro * mx % MOD;
  }
  printf("%lld %lld\n", sum, pro);
  return 0;
}

File: archived/MAY15/CHEFCK.ml
open Num;;
let xint _ = Scanf.scanf " %d" (fun x -> x);;
let xll () = Int64.of_int (xint ());;
let pm = 1000000007L;;

let mul x y m = Int64.rem (Int64.mul x y) m;;
let add x y m = Int64.rem (Int64.add x y) m;;
let min x y = if x < y then x else y;;

let () = 
  let n = xint () and k = xint () and q = xint () in
  let qa = Array.make (n + 1) pm and ms = Array.make (n + 1) pm and mp = Array.make (n + 1) pm in
  let a = xll () and b = xll () and c = xll () and d = xll () and e = xll () in
  let f = xll () and r = xll () and s = xll () and t = xll () and m = xll () in
  qa.(0) <- xll ();
  let l1 = ref (xll ()) and la = xll () and lc = xll () and lm = xll () in
  let d1 = ref (xll ()) and da = xll () and dc = xll () and dm = xll () in
  let pt = ref t in
  for i = 1 to n - 1 do
    pt := mul !pt t s;
    if !pt <= r then (
      qa.(i) <- add (mul (mul a qa.(i - 1) m) qa.(i - 1) m) (mul qa.(i - 1) b m) m;
      qa.(i) <- add qa.(i) c m;
    )
    else (
      qa.(i) <- add (mul (mul d qa.(i - 1) m) qa.(i - 1) m) (mul qa.(i - 1) e m) m;
      qa.(i) <- add qa.(i) f m;
    )
  done;
  for i = 0 to n - 1 do
    if i mod k = 0 then mp.(i) <- qa.(i)
    else mp.(i) <- min mp.(i - 1) qa.(i)
  done;
  for i = n - 1 downto 0 do
    if i mod k = k - 1 then ms.(i) <- qa.(i)
    else ms.(i) <- min ms.(i + 1) qa.(i)
  done;
  let sum = ref 0L and pro = ref 1L in
  for i = 1 to q do
    l1 := add (mul la !l1 lm) lc lm;
    d1 := add (mul da !d1 dm) dc dm;
    let ll = 1 + (Int64.to_int !l1) in
    let rr = min (ll + k - 1 + (Int64.to_int !d1)) n in
    let l = ll - 1 and r = rr - 1 in
    let pl = l / k and pr = r / k in

    let getmin x y =
      let r = ref qa.(x) in
      for i = x to y do
        r := min !r qa.(i)
      done;
      !r
    in
    let mx = (
      if pl = pr then mp.(r)
      else (
        if pl + 1 = pr then min mp.(r) ms.(l)
        else min mp.(r) (min ms.(l) mp.(k * pr - 1))
      )
    ) in
    sum := add !sum mx pm;
    pro := mul !pro mx pm;
  done;
  Printf.printf "%Ld %Ld\n" !sum !pro;